File: android/app/src/main/java/com/airbridge/app/transport/protocol/ProtocolMessage.kt
package com.airbridge.app.transport.protocol

/**
 * All message types in the AirBridge wire protocol v1.
 * Values match the Type byte in the wire format.
 * See protocol/v1/spec.md for full documentation.
 */
enum class MessageType(val value: Byte) {
    // Connection & Pairing
    HANDSHAKE(0x01),
    HANDSHAKE_ACK(0x02),
    PAIRING_REQUEST(0x03),
    PAIRING_RESPONSE(0x04),

    // File Transfer
    FILE_TRANSFER_START(0x10),
    FILE_CHUNK(0x11),
    FILE_TRANSFER_ACK(0x12),
    FILE_TRANSFER_END(0x13),

    // Screen Mirror
    MIRROR_START(0x20),
    MIRROR_FRAME(0x21),
    MIRROR_STOP(0x22),

    // Input & Clipboard
    INPUT_EVENT(0x30),
    CLIPBOARD_SYNC(0x40),

    // Keepalive
    PING(0xF0.toByte()),
    PONG(0xF1.toByte()),

    // Error
    ERROR(0xFF.toByte());

    companion object {
        fun fromByte(value: Byte): MessageType =
            entries.firstOrNull { it.value == value }
                ?: throw IllegalArgumentException("Unknown message type: 0x${value.toString(16)}")
    }
}

/**
 * A framed protocol message as read/written on the wire.
 * [payload] holds raw Protobuf bytes; callers deserialize based on [type].
 */
data class ProtocolMessage(
    val type: MessageType,
    val payload: ByteArray
) {
    companion object {
        const val PROTOCOL_VERSION = 1
        const val DEFAULT_PORT = 47821
        const val MAX_PAYLOAD_BYTES = 64 * 1024 * 1024  // 64 MB
    }

    // ByteArray requires manual equals/hashCode
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ProtocolMessage) return false
        return type == other.type && payload.contentEquals(other.payload)
    }

    override fun hashCode(): Int = 31 * type.hashCode() + payload.contentHashCode()
}

File: windows/AirBridge.Transport/Protocol/ProtocolMessage.cs
namespace AirBridge.Transport.Protocol;

/// <summary>
/// All message types in the AirBridge wire protocol v1.
/// Values match the Type byte in the wire format.
/// See protocol/v1/spec.md for full documentation.
/// </summary>
public enum MessageType : byte
{
    // Connection & Pairing
    Handshake       = 0x01,
    HandshakeAck    = 0x02,
    PairingRequest  = 0x03,
    PairingResponse = 0x04,

    // File Transfer
    FileTransferStart = 0x10,
    FileChunk         = 0x11,
    FileTransferAck   = 0x12,
    FileTransferEnd   = 0x13,

    // Screen Mirror
    MirrorStart = 0x20,
    MirrorFrame = 0x21,
    MirrorStop  = 0x22,

    // Input & Clipboard
    InputEvent     = 0x30,
    ClipboardSync  = 0x40,

    // Keepalive
    Ping = 0xF0,
    Pong = 0xF1,

    // Error
    Error = 0xFF
}

/// <summary>
/// Base representation of a framed protocol message.
/// <see cref="Payload"/> holds the raw Protobuf bytes; callers
/// deserialize to the concrete message type based on <see cref="Type"/>.
/// </summary>
public sealed record ProtocolMessage(MessageType Type, byte[] Payload)
{
    /// <summary>Protocol version this library implements.</summary>
    public const int ProtocolVersion = 1;

    /// <summary>Default TCP port for AirBridge connections.</summary>
    public const int DefaultPort = 47821;

    /// <summary>Maximum allowed payload size (64 MB).</summary>
    public const int MaxPayloadBytes = 64 * 1024 * 1024;
}
